### ISSUE TITLE
May I do some refactor? :)

### ISSUE THREAD

**Issue by Dukobpa3** [hippojay/plugin.video.plexbmc#87]
I want to divide plexbmc.py with more than 4k strings to several packages.
Will you accept my request with it?

**Comment by hippojay** [comment]
Depends what you plan on doing :)

I'm happy to accept anything that improves the plexbmc code, but i'll tell you what my overall strategy is which will help you decide what you think is worth pursuing:

Strategy - Create a single simple to use addon that replicates PHT functionality as closely as possible on the KODI platform.  It should work "out of the box" without configuration, although we will provide advanced configuration as this is an expectation of KODI power users.

To do this I was looking at the following tactics:

1. Separate display and API type code
Move API type code where we find and return code from the XML (such as movietag(), getthumb(), etc) to a separate file.  I want to be able to reuse this type of code elsewhere (see 5)

2. Move skin functions - whilst these run in the foreground, they affect background skin values and therefore I think they should be separated from the main "real-time" running code

3. Refactor the skin code.  I'm sure there are some gains to be made here.  The code is too linear for running across multiple sections and servers.  

4.Incorporate helper code into plexbmc.  Make plexbmc start the service, but also provide addon functions.  This means a single package to install and manage, which should be easier for users.  A single plexGDM module, so we don't move out of sync for changes.  Combine settings, so that it's easier to manage and share settings across the two functions (such as the issue with plex timelines)

5. Move server discovery to plexbmc helper function - so that we can do continuous discovery of servers and sections.  This in turn would mean that we can do continuous updates of skin properties to ensure that we are never waiting too long for an up to date interface.  Part of this would be intermediate caching in the helper, so the main plugin is never waiting for XML to be delivered fr the plex server.  I think doing this would have a big impact on performance and may well allow the skin to move to dynamic listitems.

I was going to start this in the following order, as I wanted to get some immediate control over the Helper and try and merge some functionality to improve the experiance:
1. Merge plexbmc helper into plexbmc - immediate control over end to end function
2. Create process to allow helper to do ongoing discovery in separate thread.(server + section)
3. Move Skin functions to separate file
4. Create process to allow helper to do continuous skin updates

Moving to separate API files and other internal refactoring was going to be left til later, as some of these architectural changes might affect what is done in this area.

One thing to bear in mind is performance.  Some of the code iterate many thousands of times and small changes can introduce larger effects.  Having helper to the heavy lifting may help smooth over a lot of these.

Does this fit with what you had in mind to work on?  I'm happy to discuss further.

**Comment by Dukobpa3** [comment]
For start I want to do just refactoring code without functionality changes.
It is needed for more flexibility.
Later we can talk about architecture of this addon.
Some like this, I think. But for now I can't see the full picture of all dependencies.
![plexbmc architecture](https://cloud.githubusercontent.com/assets/1858880/7492815/662e74c2-f404-11e4-8658-7737f64817e7.png)

**Comment by hippojay** [comment]
ok - go for it :)

**Comment by Dukobpa3** [comment]
I've done and tested first stage as I think.
![jxcva8cx0yfvxnmbvyxa 1](https://cloud.githubusercontent.com/assets/1858880/7503361/28498650-f44b-11e4-9fa9-882b33fd73b1.jpg)

Next I'll do refactor plexbmc.py to "Command" pattern.
(Empty package commands for now)

**Comment by hippojay** [comment]
Thanks.  Had a look through and changes seem reasonable.

**Comment by Dukobpa3** [comment]
```python
    if command_name is None:
        try:
            command_name = sys.argv[1]
        except:
            pass

    # todo pass parameters to constructor
    command = COMMANDS.get(command_name)()
    if command and isinstance(command, BaseCommand):
        command.execute()

    else:
        # do some
```

**Comment by hippojay** [comment]
Ok - always through the if/then was a poor way to have to organise the code.

**Comment by hippojay** [comment]
now my concerns:)
1. Have you run any of this code to see how it performs in a KODI environment?

This refactor seems (to me) to be a major piece of work.  I'm guess you work with software development as a day job?

**Comment by Dukobpa3** [comment]
> 1. Have you run any of this code to see how it performs in a KODI environment?

I'm running it on a raspberry-pi-2 with osmc + Plex on a desctop PC. It is the my master media-set.
Previous iteration have tested and merged with your latest commits.
https://github.com/Dukobpa3/plugin.video.plexbmc/tree/refactor/packages
And current - WIP.
refactor/commands

> you work with software development as a day job?

:) right

**Comment by Dukobpa3** [comment]
I've tested first set of commands. Seems like all is ok. I'm not sure that nothing is broken :) Can you help me check it?
You can get this commit.
https://github.com/Dukobpa3/plugin.video.plexbmc/commit/40bbb917d427793cacc9f43a4c8ed16fcf9445aa

Next step is commans with arguments and modes.

**Comment by hippojay** [comment]
Just found a couple of issues (two imports and one logging):
https://github.com/hippojay/plugin.video.plexbmc/commit/b3669b9629227ec24b37e996bc79ea48f59056a5

**Comment by hippojay** [comment]
Tested the following:
Amber skin functions
signin/signout/switch (I found a display issue, which isn't critical but i've commited a fix for it)
manage user
settings display
playback
transcoding playback
monitoring (plex timeline)
Channels (only apple trailed so far)

**Comment by Dukobpa3** [comment]
Good! Thanks!
I'm workin' with "skin" and "amberskin" commands for now...
Trying to untangle the pile of spaghetti :)
Will merge your fixes after finish this.

**Comment by Dukobpa3** [comment]
Done.

There is two errors in log:
1 - when the cursor gets over the "QUEUE" button
```
02:03:55 T:6000  NOTICE: commands -> amberskin : == ENTER ==
02:03:55 T:6000  NOTICE: PleXBMC.plex -> talk_to_myplex : url = https://plex.tv/pms/playlists/queue/all
02:03:55 T:6000   ERROR: C:\Users\Dukobpa3\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\packages\urllib3\util\ssl_.py:79: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
                                              InsecurePlatformWarning
02:03:55 T:6000  NOTICE: PleXBMC.plex -> talk_to_myplex : Full URL was: https://plex.tv/pms/playlists/queue/all?X-Plex-Token=CkzhRAreXKqFfok9Pn53&X-Plex-Model=unknown&X-Plex-Client-Platform=KODI&X-Plex-Client-Identifier=True&X-Plex-Device=PleXBMC&X-Plex-Product=PleXBMC&X-Plex-Platform-Version=Windows&X-Plex-Language=en&X-Plex-Version=4.0.0~beta4&X-Plex-Device-Name=My+PleXBMC+Client&X-Plex-Provides=player&X-Plex-Platform=PleXBMC
02:03:55 T:6000  NOTICE: PleXBMC.plex -> talk_to_myplex : Full header sent was: {'Connection': 'keep-alive', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'User-Agent': 'python-requests/2.6.0 CPython/2.7.8 Windows/post2012Server'}
02:03:55 T:6000  NOTICE: PleXBMC.plex -> talk_to_myplex : Full header recieved was: {'status': '200 OK', 'x-request-id': '75d872bc9ee6cc58aba21c9bca17962d', 'content-encoding': 'gzip', 'content-length': '267', 'server': 'nginx', 'connection': 'keep-alive', 'x-ua-compatible': 'IE=Edge,chrome=1', 'cache-control': 'no-cache', 'date': 'Thu, 07 May 2015 23:03:58 GMT', 'content-type': 'application/xml; charset=utf-8', 'x-rack-cache': 'miss', 'x-runtime': '0.026366'}
02:03:55 T:6000  NOTICE: PleXBMC.plex -> talk_to_myplex : ====== XML returned =======
                                            <?xml version="1.0" encoding="UTF-8"?>
                                            <MediaContainer friendlyName="myPlex" identifier="com.plexapp.plugins.myplex" machineIdentifier="b7a404fc3740b76136bd9b288f7cac64bd679247" mediaTagPrefix="/system/bundle/media/flags/" mediaTagVersion="1302321209" size="0" title1="My Queue" title2="All" viewGroup="mixed" content="mixed">
                                            </MediaContainer>
                                            ====== XML finished ======
```

2 - when tried to load channels thumbs
```
02:06:34 T:7284   ERROR: dll_fputs emulated function failed
02:06:34 T:7284   ERROR: Previous line repeats 1 times.
02:06:34 T:7284   ERROR: CCurlFile::Stat - Failed: HTTP response code said error(22) for http://192.168.56.1:32400/:/plugins/com.plexapp.system/resources/Book2.png | X-Plex-Platform=KODI&X-Plex-Client-Platform=KODI&X-Plex-User=Dukobpa3&X-Plex-Device-Name=My%20PleXBMC%20Client&X-Plex-Provides=player&X-Plex-Token=CkzhRAreXKqFfok9Pn53&X-Plex-Model=unknown&X-Plex-Platform-Version=Windows&X-Plex-Client-Identifier=True&X-Plex-Device=PleXBMC&X-Plex-Product=PleXBMC&X-Plex-Language=en&X-Plex-Version=4.0.0%7Ebeta4
```
![k38yhdny3uu7dp2vlts2a 1](https://cloud.githubusercontent.com/assets/1858880/7527710/2e96e534-f527-11e4-8edc-ce9d57b5da47.jpg)


Do you know how to fix it?

**Comment by hippojay** [comment]
1. This is an issue with the requests module, SSL and python.  It's not an error as such, but a warning about SSL connection failing sometimes (not that I have seen any).  This is only an issue for plex.tv connections at the moment (as direct connections to plex media servers are http at the moment).

2. These are working on my setup, so I'm not sure why they are not showing up for you.

**Comment by Dukobpa3** [comment]
2 - workin for me too when im not logged in plex

**Comment by hippojay** [comment]
A curl 22 is an http error code > 400.   So possibly not authorised. Where you in as a managed user?  Not sure if they work if you are not the admin.

**Comment by Dukobpa3** [comment]
Got it )))
https://github.com/Dukobpa3/plugin.video.plexbmc/commit/4cbdda3f05fd22fd793173e3d3c4731ee05e8e22#diff-c6492a453748785332285959cad8a517R425

**Comment by hippojay** [comment]
Hmm - not convinced that was the actual issue. A pipe in the url sent to KODI means that the arguments will be send as http headers rather then url arguments.    It affects all image Url's created. 

If that was the issue I would expect it to fail for me too.

**Comment by Dukobpa3** [comment]
> will be send as http headers

ahh )) ok
So it means that your "fullres_thumb" url is wrong or it is Plex bug with non-transcoded images.
I think it is Plex.
Try to set "fullres thumbnails" to true.

**Comment by Dukobpa3** [comment]
nvm. 4 a.m. I go to bed...

**Comment by Dukobpa3** [comment]
Now seems like everything is ok. I will try on clear kodi setup.

**Comment by Dukobpa3** [comment]
Do we need this command? I don't like it :)
```python
    # nt currently used
    elif command_name == "refreshplexbmc":
```

**Comment by hippojay** [comment]
Possibly not.  It was there to allow a skin refresh to take a place - but I've not used it for a long time.  It's easier to just reload the skin, which has the same affect.

It may have been used in an earlier skin (such as Quartz), but that's a dead skin.  I think it could probably be removed now.

**Comment by Dukobpa3** [comment]
I've done with all "script" commands.
Help me check it please.
You should remove previous version before update . I don't know how to clear all automatically.

Next is - "plugin" requests. I need few time for investigate best solution for it.

**Comment by Dukobpa3** [comment]
One more thing.
I've investigated sources of some other plugins from top ratings.
No one is have script and pluginsource processor in one addon.
I think we should make package of addons for different tasks eg skin settings, content providing, remote content playing, etc.

Also I want move it to official repo ;)
What you think about it?
You did a great job and I think the results should be available to the public.

**Comment by wickning1** [comment]
As far as I know adding it to the official repo won't be allowed by the Kodi developers.  It's been discussed multiple times in the past. 

If it's technically possible, and it appears that it is, I think we have a lot to gain by putting the remote playing helper in with the main app.  They'll be able to share code easily without ending up with weird dependency problems (dependency can only go in one direction), and they can communicate with shared memory instead of having to serialize objects into properties on a window object, or communicate through the jsonrpc or network stack.  I'm guessing those benefits outweigh the idea of being a little better organized.

What skin settings are you referring to as warranting a new plugin?

**Comment by Dukobpa3** [comment]
Maybe it would be one addon as you explained before. But be able to disable each scoupe of functionality.
Eg I don't need remote playing.
Or someone don't want to use skin modifications.

**Comment by wickning1** [comment]
I'm sure we'll add a setting to turn helper off.  It's only been in the main project for about a week.

**Comment by Dukobpa3** [comment]
> won't be allowed by the Kodi developers. 

Can you explain reasons of Kodi team?
I've investigated requirements and can't find something prohibited in PleXBMC.

> What skin settings are you referring to as warranting a new plugin?

After refactor I've understood that much code is shelfs and other skin things.
By default content provider don't provide this functionality.

**Comment by Dukobpa3** [comment]
> They'll be able to share code easily without ending up with weird dependency problems

It can be some "module" type addon.

**Comment by hippojay** [comment]
OK - my thoughts:

dukobpa3:
> No one is have script and pluginsource processor in one addon.

Not many, if any, do - you are right.  This is probably because other people don't need to do this.    

The problem I'm trying to solve is one of persistence.    As you've seen, addons are stateless - so i have this save/load cycle every time.  Before I implemented caching, I was having to discover on every cycle to, which was painful!  A script/service in KODI can solve this.

Now, we could implement as a separate addon (as we do now), but we come up against the issue of code updates, code sharing and usability.    A bigger problem is that users then have to load two addons to get the full feature set, have to configure both and in some cases need to ensure settings are similar.

A single package just seems to make more sense.  Plus, as i add more advanced features, I come up against more issues that require some overarching service that runs (such as playlist items - easy to implement, hard to perfect with plex timeline reporting).

dukobpa3:
>Also I want move it to official repo ;)
> Can you explain reasons of Kodi team?
> I've investigated requirements and can't find something prohibited in PleXBMC.

This is (or perhaps was) the unwritten rule :)  And has some small history.  Waaaay back before Plex was Plex, it was part of XBMC (part of the effort to get XBMC working on OSX).  There was a falling out between Team XBMC and this OSX port developer, and plex was born as a fork of XBMC.   Team XBMC didn't take to kindly and ever since then there has been a degree of bad blood between the two.

I originally approached the team back in 2011 to get this on the official repo and was told by two members of Team XBMC that they would veto any attempts to add this to the official repo.  (So all team memebers weren't against this, just two).  Therefore I took it no further.  

However, since then I've been told that these two members have now left, so now there is a better possibility. The only thing that has stopped me approaching them again was a) the plugin wasn't in a fully featured state and b) he code probably wouldn't meet the design requirements.  a) is now fixed and b) is what you are doing - so it is probably worth a try again.  I would need to check with Plex team to make sure that we weren't infringing their fair use trademark rules (which are posted on the plex forums).

dukobpa3:
> After refactor I've understood that much code is shelfs and other skin things.
> By default content provider don't provide this functionality.
and
>Eg I don't need remote playing.
>Or someone don't want to use skin modifications.

You are right, but I think the convieniance of having it packaged together is better than splitting it up and making it more difficult to bring together.    Having settings to disable, or perhaps some smart detection (i.e. if the skin isn't plexbmc compatible, then switch off skin updates)

In a nutshell, I think we need a persistent service running with PleXBMC that can provide all the background stuff that a pluginsource addon can't.  It's the next step in providing a full client interface.

>You did a great job and I think the results should be available to the public.

Thanks - I appreciate that and also the work you've done on the refactor.  It makes it more professional, even if it makes it more difficult for me to work out whats going on!  But i'm always up for learning better ways to architect software.

So - my proposed next steps would be:
* Make a v4 release (non-refactor) based on current code
* Merge in refactor and test, test, test.
* Look to release this as interim v4.1/4.5 version

In parallel on the refactored code, I want to look to merge in plexbmc helper, and use common code base for it (plexgdm, plex_server).  

any thought?

**Comment by Dukobpa3** [comment]
> A single package just seems to make more sense. 

Ok, I got it. But we should keep internal architecture as a couple of modules instead of monolithic plate of spaghetti.

> There was a falling out between Team XBMC and this OSX port developer

:) Opensorce guys vs proprietary guys conflict :) Ok, thx for explaining it.

> it makes it more difficult for me to work out whats going on

I will add more dockstrings later. I'm sure you will understand quickly with this all.
eg. if you want to modify some command processor you don't need to read 4.5K strings. Instead of this  you are just open some "command" class and fix it as you need. Command and Strategy is a very useful patterns for stateless workflow.

**Comment by hippojay** [comment]
It's just a learning curve that's all :-) I knew where everything was in the old code, but that doesn't mean to say that other people did!  I've never referred to my code as anything other than ugly - it's just something I never had the time to sort out. So thanks for doing that.

**Comment by Dukobpa3** [comment]
Hi. I've pushed current result to "/refactor/script-commands" branch.
https://github.com/Dukobpa3/plugin.video.plexbmc/tree/refactor/script-commands

You can grab and check it.
I think I should to continue with "pluginsourse" commands in another branch due to complexity of this part.

**Comment by hippojay** [comment]
Just had a look and I'm not sure what you are doing with the split between pluginsource and script?

Are you trying to split the script type commands (runscript) from the plugin type command operations (plugin://).  I'm not sure this is the right split as it will break the interface.  i might have the wrong end of the stick though.

**Comment by Dukobpa3** [comment]
>> Are you trying to split the script type commands (runscript) from the plugin type command operations (plugin://).

Yup. Both of this have a different interface and splitting would provide simpler code as start point.
Fo now we are have much cross-dependencies due to current solution.
Splitting will be just internal.

**Comment by Dukobpa3** [comment]
Also you are have this splitting already but with "if... else"
```
    # else move to the main code
    else:
        # Here is command operations (plugin://)
```
I'm just want to do it more gracefully

**Comment by hippojay** [comment]
ok.  no problem

**Comment by Dukobpa3** [comment]
```python
    args = sys.argv[2:]
    command = COMMANDS.get(command_name)
    if command and issubclass(command, BaseCommand):
        printDebug.debug("executing command: %s; with args: %s" % (command, args))
        command(args).execute()
```
Script command are running from system arguments. And this part is all we need to launch some actions.

For plugun request we should to parse incoming url. It is much different logic.

**Comment by hippojay** [comment]
Had a quick run through and all looks ok.  Had one issue with commands_amberskin.py where you missed a GlobalEnum when you were removing them (line 629).  Can't do a fix cos my git is a bit out of sync at the moment and don;t have time to start over.

**Comment by Dukobpa3** [comment]
You are shouldn't to fix any. Just grab last commit from this branch:
https://github.com/Dukobpa3/plugin.video.plexbmc/tree/refactor/script-commands

**Comment by hippojay** [comment]
Ok - i've found a problem, which I'm not able to work out how to fix.  Bear with me because it's not easy to explain, but hopefully you'll understand.

First off - audio is missing from baseCommands.  Easy enough to add, which I have done in my local copy

Next , There is an issue with the basecommands class, which is stopping the context menu entries form work (but I daresay a lot of other stuff too).  If you want to reproduce the error, then right-click any media file bring displayed and click on "select subtitles".  You will get a tuple self.arg[1] out of range error.

The arguments passed to baseCommand are as a list.  baseCommands will take these as a *args and so they will be packed in a tuple.  So the entire list will be as arg[0].  Howeverm, do what I might, I cannot get anything to access the arg[1] index.  As a test I've tried assigning arg[0] to a variable, but this still won;t let me access [1].  I've tried adding a * to the input args in plexbmc.py (so that the list is converted to a tuple) but this doesn;t work either.

No matter what I do, I cannot get the list/tuple to access [1].  It's as if it is a string rather than a list/tuple, but running a type () on it confirms it as a tuple.  

Could you have a look and let me know where I've gone wrong.  I can only assume it's something straight forward I;ve overlooked.

**Comment by hippojay** [comment]
Hmm - fixed it by doing this.  Would be good to hear your thoughts:
https://github.com/hippojay/plugin.video.plexbmc/commit/36ba0a9e9a156fc28c429a459f3be54ffec44de3

**Comment by Dukobpa3** [comment]
It's ok, *args should be passed with asterisk.
I've got a typo.

**Comment by Dukobpa3** [comment]
I've merged your fix. Thx.

**Comment by Dukobpa3** [comment]
Look at this realization for Emby (MediaBrowser).
```
Advantages of new AddOn:
- Works with native skins
- Much faster library browsing
- Native sorting/deleting/resume
- Native view setting support
```
https://github.com/MediaBrowser/Emby.Kodi

I think we would to keep in mind this ideas.

**Comment by hippojay** [comment]
It's something I have looked at before, but I also see issues:
* Support for separate libraries (for example, all movie sections would need to be merged into a single movies library)
* Support for user authentication (plexHome and possible downstream changes to authentication)

I've looked at the idea of syncing before (probably 3 years ago), but my concerns have always been time taken to do it and the lag between the KODI view and the Plex view (for thing such as resuming media and watch status).    How often to refresh and does the refresh have an impact on the foreground performance.

There are benefits of skin integration and use of standard widgets - which is why I've not ruled it out - but it's not my priority.  I'm letting the Emby team work out the kinks first and then i'll have a look.  If someone else wants to pick this up, then I would be happy to support them.

**Comment by Dukobpa3** [comment]
https://github.com/MediaBrowser/Emby.Kodi/blob/master/README.md

>
#### Important note about user collections/nodes

> Emby has the ability to create custom nodes/folders for your Media, such as having a seperate folder for your "Kids Movies" etc. In Kodi this isn't supported, you just have "movies" or "tvshows". But... Kodi let's you create your own playlists and tags to get this same experience. During the sync the foldernode from the Emby server is added to the movies that are imported. In Kodi you can browse to Movie library --> tags and you will have a filtered result that points at your custom node. If you have a skin that let's you create any kind of shortcut on the homescreen you can simply create a shortcut to that tag. Another possibility is to create a "smart playlist" with Kodi and set it to show the content of a certain tag.

**Comment by hippojay** [comment]
Yeah. I think I've posting in the plex thread about using tags to simulate library sections.  Like I said, I'm not ruling it but need to be convinced.

**Comment by Dukobpa3** [comment]
I think that using of native Kodi library is much better solution than provide some non-obvious mechanics.

But we are need to have well-functioning content provider for Plex-server in any case.
It means plex-home, queue and other Plex features.

**Comment by hippojay** [comment]
There are compromises to be made on both paths - although Emby own the client and server code, so they have an advantage of knowing what is coming up and the impact it will have, or altering the server to work with the client better.

We could probably get a POC up in a week to see what it would look like.

**Comment by Dukobpa3** [comment]
> POC
?

**Comment by hippojay** [comment]
Proof of concept :)

**Comment by Dukobpa3** [comment]
ahh))

**Comment by Dukobpa3** [comment]
I could not get to work this Emby addon.

**Comment by hippojay** [comment]
With emby or were you trying to alter for plex?

I've not tried it yet, as it means setting up a separate emby server.

**Comment by Dukobpa3** [comment]
With Emby

**Comment by hippojay** [comment]
Hows the code btw.  I'm kinda holding back a little because I don't want to add stuff to the current code and have to re-implement for any refactored code.  I'm away for a bit anyway, but I can see a number of things that need some attention when i get back.  Thanks

**Comment by Dukobpa3** [comment]
You can do what you want with your code. I will implement your fixes in refactored version myself, don't worry about it.
I'm investigating other realizations for now so I'm have pause with production for now. I'm trying it on my copies of Kodi.
I see that we have different views on some points about "how it would be". I think I should wait few time for your POC described :)

**Comment by Dukobpa3** [comment]
My point is (sorted by priority):
1. Well-functioning content-provider (pluginsource)
 - with authorization
 - with plex home
 - with other features like a last music updates

2. All the rest
 - skins support (At this point we are should investigate skins that implements Plex support, eg Titan or other MediaBrowser skins because of MediaBrowser have the same architecture as Your Addon)
 - watchdog for plex monitoring

**Comment by hippojay** [comment]
That's ok. When I get back I'll start to add some of the requested features and fixes - including a couple of yours :-)

**Comment by Dukobpa3** [comment]
Hi guys.
Sorry, but I've gone to emby server from plex.
You can use current results of my work "as is" if you like it. It is contains total refactoringt "script" commands.
Also I can help you with any your questions and tasks.
Thanks for teamwork :)

**Comment by Rockycjc** [comment]
This probably isn't the forum for that discussion but may I suggest finding a place for it and posting your reasoning.  I'd be interested in emby if it is a better solution than PLEX.

**Comment by Dukobpa3** [comment]
Better just for me. It is have many disadvantages also.
